src/webots/nodes/WbShape.cpp

```diff
@@ -387,9 +390,12 @@ void WbShape::pickColor(WbRgb &pickedColor, const WbRay &ray, double *roughness,
       return;  // default value
 
     // combine colors
-    pickedColor.setRed(diffuseColor.red() * textureColor.red() * paintColor.red());
-    pickedColor.setGreen(diffuseColor.green() * textureColor.green() * paintColor.green());
-    pickedColor.setBlue(diffuseColor.blue() * textureColor.blue() * paintColor.blue());
+    pickedColor.setRed((1 - paintContribution) * diffuseColor.red() * textureColor.red() +
```

Comment:
I would use a float value in these instructions so that the expected type is set (sometimes this could lead to errors):
```suggestion
    pickedColor.setRed((1.0f - paintContribution) * diffuseColor.red() * textureColor.red() +
```